### PR TITLE
fix poly degree 0 underflow

### PR DESCRIPTION
**Pull request by smtmfft**
substract underflow if poly's degree is 0, which happens if we have an empty/constant value gate.

Another issue is if I don't update the rust-toolchain to 1.66, I got:
```
error: package `constant_time_eq v0.3.0` cannot be built because it requires rustc 1.66.0 or newer, while the currently active rustc version is 1.64.0
```

### CODE REVIEW

File: halo2_proofs/src/plonk/evaluation.rs [hunk 1]
```diff
@@ -976,7 +975,11 @@ impl<C: CurveAffine> Evaluator<C> {
...
     fn compute_cluster_idx(degree: usize, max_cluster_idx: usize) -> usize {
-        let mut idx = (31 - (degree as u32).leading_zeros()) as usize;
+        let mut idx = if degree != 0 {
+            (31 - (degree as u32).leading_zeros()) as usize
+        } else {
+            0
+        };
         if 1 << idx < degree {
             idx = idx + 1;
         }
```
Comment (einar-taiko) [inline halo2_proofs/src/plonk/evaluation.rs]: I suggest the following rewrite:

```rust
let mut idx = 0;

if degree != 0 {
    idx = degree.ilog2() as usize;

    if !degree.is_power_of_two() {
         idx += 1;
    }   
} 
   
```   
See https://godbolt.org/z/YEPd6n5cE

File: halo2_proofs/src/plonk/evaluation.rs [hunk 2]
```diff
@@ -1245,3 +1248,13 @@ pub fn evaluate<F: Field, B: Basis>(
     });
     values
 }
+
+mod test {
+    #[test]
+    #[should_panic(expected = "attempt to subtract with overflow")]
+    fn compute_cluster_idx_underflow() {
```
Comment (einar-taiko) [inline halo2_proofs/src/plonk/evaluation.rs:1259]: Nice!

File: halo2_proofs/src/fft/parallel.rs [hunk 3]
```diff
@@ -120,7 +120,7 @@ fn split_radix_fft<Scalar: Field, G: FftGroup<Scalar>>(
     let tmp_filler_val = tmp[0];
     let mut t1 = vec![tmp_filler_val; split_m];
     for i in 0..split_m {
-        t1[arithmetic::bitreverse(i, log_split)] = a[(i * sub_n + sub_fft_offset)];
+        t1[arithmetic::bitreverse(i, log_split)] = a[i * sub_n + sub_fft_offset];
```
Comment (einar-taiko) [inline halo2_proofs/src/fft/parallel.rs:123]: Reformat the whole thing or leave this out.

File: halo2_proofs/src/plonk/evaluation.rs [hunk 4]
```diff
@@ -434,7 +434,7 @@ impl<C: CurveAffine> Evaluator<C> {
             .enumerate()
             .take(num_clusters)
         {
-             cluster.resize(1 << cluster_idx, domain.empty_lagrange());
+            cluster.resize(1 << cluster_idx, domain.empty_lagrange());
```
Comment (einar-taiko) [inline halo2_proofs/src/plonk/evaluation.rs:437]: Reformat the whole thing or leave this out.

File: rust-toolchain [hunk 5]
```diff
@@ -1 +1 @@
-1.64.0
+1.66.0
```
Comment (einar-taiko) [inline rust-toolchain:1]: This is in sync with upstream https://github.com/privacy-scaling-explorations/halo2/blob/main/rust-toolchain#L1 :+1:

**Comment by einar-taiko** [comment]
Great work! I specially like the test case :dizzy:

**Review by einar-taiko — APPROVED** [review]
The fix is correct.

Please treat my comments as suggestions.

File: halo2_proofs/src/plonk/evaluation.rs (hunk 4 above)
Comment (smtmfft) [reply halo2_proofs/src/plonk/evaluation.rs:437]: yes, make a global cargo fmt.

File: halo2_proofs/src/plonk/evaluation.rs (hunk 1 above)
Comment (smtmfft) [reply halo2_proofs/src/plonk/evaluation.rs]: ilog2() is unstable, use log2().floor() instead.

File: halo2_proofs/src/plonk/evaluation.rs [hunk 6]
```diff
@@ -975,14 +975,15 @@ impl<C: CurveAffine> Evaluator<C> {
...
-            (31 - (degree as u32).leading_zeros()) as usize
-        } else {
-            0
-        };
-        if 1 << idx < degree {
-            idx = idx + 1;
+        let mut idx = 0;
+        if degree != 0 {
+            // same as unstable degree.ilog()
+            idx = (degree as f32).log2().floor() as usize;
```
Comment (einar-taiko) [inline halo2_proofs/src/plonk/evaluation.rs]: Good catch with the versioning.

I would prefer `idx = (31 - (degree as u32).leading_zeros()) as usize` over introducing floating-point math.

**Review by einar-taiko — APPROVED** [review]
What do you think about running `cargo clippy --fix` on the whole thing as well?

File: halo2_proofs/src/plonk/evaluation.rs (hunk 6 above)
Comment (smtmfft) [reply halo2_proofs/src/plonk/evaluation.rs]: NP, let's back to the origin code :)

**Comment by smtmfft** [comment]
> What do you think about running `cargo clippy --fix` on the whole thing as well?

I guess it should be ok, let me try

**Comment by smtmfft** [comment]
All CI & lint issues are addressed!

**Pull request merged**